rfd/0003-plugins-unknown-events.md

```diff
@@ -0,0 +1,352 @@
+  // session_id is the ID for a given session in an UUIDv4 format.
+  string session_id = 1;
+  // StartIndex is the index of the event to resume the stream after.
+  // A start_index of 0 creates a new stream.
+  int32 start_index = 2;
+}
+
+// Order specifies any ordering of some objects as returned in regards to some aspect
+// of said objects which may be trivially ordered such as a timestamp.
+enum Order {
+  ORDER_DESCENDING_UNSPECIFIED = 0;
+  ORDER_ASCENDING = 1;
+}
+
+// GetUnstructuredEventsRequest is a request with the needed data to fetch events.
+message GetUnstructuredEventsRequest {
+  // namespace, if not set, defaults to 'default'
+  string namespace = 1;
+  // start_date is the oldest date of returned events
+  google.protobuf.Timestamp start_date = 2;
+  // end_date is the newest date of returned events
+  google.protobuf.Timestamp end_date = 3;
+  // event_types is optional, if not set, returns all events
+  repeated string event_types = 4;
+  // limit is the maximum amount of events returned
+  int32 limit = 5;
+  // start_key is used to resume a query in order to enable pagination.
+  // If the previous response had LastKey set then this should be
+  // set to its value. Otherwise leave empty.
+  string start_key = 6;
+  // order specifies an ascending or descending order of events.
+  // A value of 0 means a descending order and a value of 1 means an ascending order.
+  Order order = 7;
+}
+
+// EventsUnstructured represents a list of events.AuditEvent in unstructured format.
+message EventsUnstructured {
+  // Items is a list of unstructured formatted audit events.
+  repeated EventUnstructured items = 1;
+  // the key of the last event if the returned set did not contain all events found i.e limit <
+  // actual amount. this is the key clients can supply in another API request to continue fetching
+  // events from the previous last position
+  string last_key = 2;
+}
+
+// EventUnstructured represents a siggle events.AuditEvent in an unstructured format.
+message EventUnstructured {
+  // type is the type of the event.
+  string type = 1;
+  // id is the unique ID of the event.
+  // If the underlying event defines an ID, it will be used, otherwise
+  // it is a SHA256 hash of the event payload.
+  string id = 2;
+  // time is the time when the event was generated.
+  google.protobuf.Timestamp time = 3;
+  // index is the index of the event.
+  int64 index = 4;
+  // unstructured is the unstructured representation of the event payload.
+  google.protobuf.Struct unstructured = 5;
+}
+
+```
+
+Auth server converts the events to `structpb.Struct` and returns them using `EventUnstructured` struct.
+The conversion is done using the default event JSON representation for each event type
+besides the `print` event type. For the `print` event type, an extra field
+`data` is appended to the `structpb.Struct` because the default JSON representation does not include the
+`data` field which is required by the event handler.
+
```

Comment:
How does this encoding process look on the `auth` side? I am imagining something that is the moral equivalent of this :

```go
func MarshalEvent(structuredEvent *someTeleportEventType) *EventUnstructured {
  jsonText, _ := json.Marshal(&structuredEvent)
  
  var structMap map[string]any
  _ = json.Unmarshal(jsonText, &structMap)

  return &EventUnstructured {
    unstructured: structpb.NewStruct(structMap),
  }
}
```
(although hopefully without necessarily having to go all the way down to text to perform the `eventType` to `map[string]any` conversion).

Or have I got the wrong end of the stick entirely?

Reply:
The process is similar but you can unmarshal the JSON directly into the struct ([real code](https://github.com/gravitational/teleport/pull/27026/files#diff-0d992726e983ecd7ccd935e9769652763907ad1a88ae462050d2770e557d3747R5224-R5257)):

```go

func MarshalEvent(structuredEvent *someTeleportEventType) *EventUnstructured {
  jsonText, _ := json.Marshal(&structuredEvent)
  
  var struct structpb.Struct
  _ = struct.UnmarshalJSON(jsonText)

  return &EventUnstructured {
    unstructured: struct,
  }
}


```